### PR TITLE
Allow functors to have non-legacy mode input and output

### PR DESCRIPTION
**Pull request by riaqn**
Based on #5112 

This PR adds the support for functors to have input and output of non-legacy modes. Several things that are missing are marked as `CR-soon`.

Strongly recommend to review by commits.

### CODE REVIEW

File: typing/env.ml [hunk 1]
```diff
@@ -776,7 +776,13 @@ let clda_mode = Mode.Value.(
...
       ~hint_comonadic:Class_legacy_comonadic)
 
-let fcomp_res_mode = Types.functor_res_mode |> Mode.Alloc.disallow_right
+(** In this file, a functor's return is only used to access the types inside
+(such as F(M).t). Therefore, the return having the weakest mode is sufficient.
+*)
+let fcomp_res_mode = Mode.Value.(max |> disallow_right)
+
+(** Accessing `F(M).t` is not closing over anything *)
+let fcomp_res_mode_with_locks = (fcomp_res_mode, locks_empty)
```
Comment (dkalinichenko-js) [inline typing/env.ml:779]: Let's add tests for accessing types inside functor returns.

File: typing/includemod_errorprinter.ml [hunk 2]
```diff
@@ -628,22 +647,19 @@ module Functor_suberror = struct
...
     let single_diff ~is_modal g e more =
-      let _arg, mty, (mode1, locks) = g.With_shorthand.item in
-      let mode2 =
-        Types.functor_param_mode
-        |> Mode.alloc_as_value
-        |> Mode.Value.disallow_left
+      let _arg, mty1, mode1_with_locks = g.With_shorthand.item in
+      let mty2, mode2 =
+        match e.With_shorthand.item with
+        | Types.Unit -> Format.dprintf "()", Mode.Alloc.legacy
```
Comment (dkalinichenko-js) [inline typing/includemod_errorprinter.ml]: Seems bad to hardcode `Mode.Alloc.legacy` here, maybe let's factor this into a helper definition?

File: typing/printtyp.ml [hunk 3]
```diff
@@ -884,6 +884,15 @@ let set_printing_env env =
     printing_cont := [cont];
   end
 
+(* CR-soon zqian: Currently we immediately backtrack each mutation, which might
+cause incoherent types/modes in a single printing. Instead, we should move
```
Comment (dkalinichenko-js) [inline typing/printtyp.ml:888]: This issue was already present before this PR, but if you do have a test on hand that demonstrates it, could you add it? If you don't, feel free to resolve this.

File: typing/typemod.ml [hunk 4]
```diff
@@ -158,6 +158,55 @@ let rec path_concat head p =
   | Papply _ -> assert false
   | Pextra_ty (p, extra) -> Pextra_ty (path_concat head p, extra)
 
+let infer_modalities ~loc ~env ~md_mode ~mode =
+    (* Values are packed into a structure at modes weaker than they actually
+      are. This is to allow our legacy zapping behavior. For example:
```
Comment (dkalinichenko-js) [inline typing/typemod.ml:163]: This is an already existing comment, but could you expand on how this helps with the legacy zapping behavior?

File: typing/typemod.ml [hunk 5]
```diff
@@ -158,6 +158,55 @@ let rec path_concat head p =
...
+
+      module M = struct
+        let foo x = x
+        let bar = use_portable foo
+      end
+      module type S = module type of M
```
Comment (dkalinichenko-js) [inline typing/typemod.ml:169]: What role does `module type S = module type of M` serve in this example?

File: typing/typemod.ml [hunk 6]
```diff
@@ -158,6 +158,55 @@ let rec path_concat head p =
...
+    let mode, _ = Mode.Value.newvar_above mode in
+    (* Upon construction, for comonadic (prescriptive) axes, module
+    must be weaker than the values therein, for otherwise operations
+    would be allowed to performed on the module (and extended to the
+    values) that's disallowed for the values.
+
+    For monadic (descriptive) axes, the restriction is not on the
+    construction but on the projection, which is modelled by the
+    [Diff] modality in [mode.ml]. *)
+    begin match Mode.Value.Comonadic.submode
```
Comment (dkalinichenko-js) [inline typing/typemod.ml:190]: Also existing code, but could we move this check into Mode.Modality.infer?

File: typing/typemod.ml [hunk 7]
```diff
@@ -1996,8 +2040,16 @@ and transl_signature env {psg_items; psg_modalities; psg_loc} =
       match sincl.pincl_kind with
       | Functor ->
         Language_extension.assert_enabled ~loc Include_functor ();
-        let sg, incl_kind =
-          extract_sig_functor_open false env smty.pmty_loc mty sig_acc
+        let sg, mode, incl_kind =
+          extract_sig_functor_open false env smty.pmty_loc mty sig_acc md_mode
+        in
+        let zap_modality =
+          Ctype.zap_modalities_to_floor_if_modes_enabled_at Alpha
```
Comment (dkalinichenko-js) [inline typing/typemod.ml:2054]: Why this behavior?

Comment (dkalinichenko-js) [reply typing/typemod.ml:2054]: Could you add tests for the difference between `Stable` and `Alpha`?

File: typing/typemod.ml [hunk 8]
```diff
@@ -2497,32 +2549,37 @@ let path_of_module mexp =
 (* Check that all core type schemes in a structure
    do not contain non-generalized type variable *)
 
-let rec nongen_modtype env f = function
+let rec nongen_modtype env f g = function
```
Comment (dkalinichenko-js) [inline typing/typemod.ml:2560]: Could you add a comment describing what `f` and `g` do? (or perhaps give them descriptive names)

File: testsuite/tests/parsetree/source_jane_street.ml [hunk 9]
```diff
@@ -753,10 +745,7 @@ module F : functor () -> sig end @@ stateless
 module (G @ portable) () = F
 
 [%%expect{|
-Line 1, characters 27-28:
-1 | module (G @ portable) () = F
-                               ^
-Error: This is "contended", but expected to be "uncontended" because it is a functor body.
+module G : functor () -> (functor () -> sig end) @ contended @@ portable
```
Comment (dkalinichenko-js) [inline testsuite/tests/parsetree/source_jane_street.ml:749]: Could you add a syntax example for a functor accepting another functor as an argument, both annotated with different modes?

File: testsuite/tests/typing-modes/functor.ml [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+end
+[%%expect{|
+module type S = sig val f : unit -> unit end
+|}]
+
+module type T = sig
+    val g : unit -> unit
+end
+
+(* functor parameter's unspecified mode axes default to legacy *)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/functor.ml:22]: Could you add an `S -> T = functor ...` or other similar example here? Also, examples with multiple arguments, including annotations on no arguments, only on the first argument and only on the second argument.

File: testsuite/tests/typing-modes/functor.ml [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+[%%expect{|
+Line 2, characters 8-9:
+2 |     let g = M.f
+            ^
+Error: This is "nonportable", but expected to be "portable" because it is inside a "portable" structure.
+|}]
+
+(* At top-level, functors, just like functions, will have their parameter mode and return
+mode zapped to legacy. Therefore, we explicitly annotate the functor type. Alternatively,
+put the functor at a deeper level, and its return mode can be infered. *)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/functor.ml]: What does putting the functor at a deeper level mean here? could you give an example?

File: testsuite/tests/typing-modes/functor.ml [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+[%%expect{|
+module M :
+  sig
+    val f : unit -> unit @@ stateless
+    val f' : unit -> unit
+    val g : unit -> unit @@ portable
+  end
+|}]
+
+(* CR-soon zqian: the inner functor should be forced to once as well*)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/functor.ml:301]: Could you expand this example? Is this unsound and why?

File: testsuite/tests/typing-modes/functor.ml [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+[%%expect{|
+|}]
+
+(* CR-soon zqian: the inner functor should be forced to once as well *)
+module F (M : S @ once) () = struct end
+[%%expect{|
+module F : functor (M : S @ once) () -> sig end @@ stateless
+|}]
+
+(* CR-soon zqian: the inner functor should be forced to once as well *)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/functor.ml:314]: Is `once` intentional here?

File: testsuite/tests/typing-modes/functor.ml
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/functor.ml:1]: Could you add:
1. More tests for monadic modalities like `contended`
2. Tests for `once` and `unique` with actual items in the signature
3. Tests for `local`
4. Tests for multiple modalities combined, incl both monadic and comonadic
5. Defining a functor inside a functor
6. A functor accepting another functor as an argument

File: typing/includemod_errorprinter.ml (hunk 2 above)
Comment (riaqn) [reply typing/includemod_errorprinter.ml]: The value here doesn't matter, because `Types.Unit` doesn't have modes on it. I've rewritten the code so it's less confusing. Commit "simplify [single_diff]".

File: typing/env.ml (hunk 1 above)
Comment (riaqn) [reply typing/env.ml:779]: added in commit "add test about F(M).t".

File: typing/printtyp.ml (hunk 3 above)
Comment (riaqn) [reply typing/printtyp.ml:888]: Sadly I don't have an example on hand.

File: typing/typemod.ml (hunk 5 above)
Comment (riaqn) [reply typing/typemod.ml:169]: explained in the comments

File: typing/typemod.ml (hunk 6 above)
Comment (riaqn) [reply typing/typemod.ml:190]: the module Mode is intended as mostly algebra and doesn't concern about ocaml specifics. Therefore, I think it's better to have this detail in `typemod`.

File: typing/typemod.ml (hunk 7 above)
Comment (riaqn) [reply typing/typemod.ml:2054]: As the name suggests, when `mode_alpha` is given, we zap modalities to strongest instead of legacy, which potentailly breaks legacy code.

>Could you add tests for the difference between Stable and Alpha?

I think it would be hard to demonstrate the difference in a single file. I think really, the thing we want to ensure is that:
- under `mode = stable` (the default), all legacy code builds without issue
- Under `mode = alpha` (used by some of our internal code base), the type checking behaves as we desired.

The first is tested by the fact that all other tests pass. The second is tested by files in `typing-modes`. Therefore, I suggest we keep the current way.

File: testsuite/tests/typing-modes/functor.ml (hunk 13 above)
Comment (dkalinichenko-js) [reply testsuite/tests/typing-modes/functor.ml:314]: Ah, I see that this is a capture.

File: testsuite/tests/typing-modes/functor.ml (hunk 12 above)
Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml:301]: Yeah this is unsound. This will be fixed by another PR soon.
The unsoudness exists before, so unrelated to the current PR.

File: testsuite/tests/typing-modes/functor.ml (hunk 11 above)
Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml]: I rewrote this and added example in commit "improve tests".

File: testsuite/tests/typing-modes/functor.ml (hunk 10 above)
Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml:22]: done in commit "improve tests".

File: testsuite/tests/parsetree/source_jane_street.ml (hunk 9 above)
Comment (riaqn) [reply testsuite/tests/parsetree/source_jane_street.ml:749]: done in commit "improve tests".

File: testsuite/tests/typing-modes/functor.ml
Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml:1]: wrt 2: Can you say a bit more about `once` and `unique`? Why this specific two?

Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml:1]: wrt 4: do you mean modes, or modalities? Note that this PR is about the modes on functor arrow types, so not directly related to modalities.

Comment (dkalinichenko-js) [reply testsuite/tests/typing-modes/functor.ml:1]: 4: oops, modes, yes

Comment (dkalinichenko-js) [reply testsuite/tests/typing-modes/functor.ml:1]: 2: i think what i wanted to test here is actually tested by your unsound example, so disregard this comment. probably still good to add more tests with non-empty signatures and structures just in case

Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml:1]: points 3, 5, 6 is addressed by commit "add tests for suggestion 3 5 6".

Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml:1]: wrt 1 and 4 - note that this PR doesn't introduce any logic that treats monadic and comonadic axes differently - do you have some particular concerns in mind?

File: testsuite/tests/typing-modes/functor.ml [hunk 14]
```diff
@@ -346,3 +346,30 @@ module F (M : Portable_Portable) = (M : Portable_Nonportable)
 module F : functor (M : Portable_Portable) -> Portable_Nonportable @@
   stateless
 |}]
+
+(* refering to [F(M).t] is allowed even if [M] is weaker than what [F] wants *)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/functor.ml:516]: Should we allow this behavior? I think it is sound, since `F(M).t` can itself only refer to type, not value, declarations of `M`, which are not affected by modes. It might still be good to forbid this to avoid confusion.

File: testsuite/tests/typing-modes/functor.ml
Comment (dkalinichenko-js) [reply testsuite/tests/typing-modes/functor.ml:1]: No, just making sure it does treat monadic and comonadic axes the same.

File: testsuite/tests/typing-modes/functor.ml [hunk 15]
```diff
@@ -441,3 +441,24 @@ module F = F @@ stateless nonportable
...
+(* CR-soon zqian: the error message should refer to [M.f] as the violation. *)
+module F (M : S @ local) : S @ global = struct include M end
+[%%expect{|
+Line 1, characters 47-56:
+1 | module F (M : S @ local) : S @ global = struct include M end
+                                                   ^^^^^^^^^
+Error: This is "local", but expected to be "global" because it is inside a structure.
+|}]
+
+(* some higher order functor *)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/functor.ml:553]: Can you add tests making sure that variance works correctly for higher-order functors? E.g., add another example like `F(G)` below, but with `G' : S @ m1 -> ...` and add cases for `G'(...)` both succeeding and failing.

File: testsuite/tests/typing-modes/functor.ml [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/functor.ml:553]: Can you add tests for functor type inclusion between higher order functors?

File: testsuite/tests/typing-modes/functor.ml [hunk 17]
```diff
@@ -0,0 +1,464 @@
...
+module F : functor (M : S @ once) () -> sig end @@ stateless
+|}]
+
+(* CR-soon zqian: the inner functor should be forced to once as well *)
+module F (M : S @ unique) () = struct end
+[%%expect{|
+module F : functor (M : S @ unique) () -> sig end @@ stateless
+|}]
+
+(* testing functor type inclusion *)
```
Comment (dkalinichenko-js) [inline testsuite/tests/typing-modes/functor.ml:320]: Could you add a few tests for functor type inclusion with intermediate values on mode axes? (e.g. testing `nonportable < shareable < portable` for both covariant and contravariant positions)

File: testsuite/tests/typing-modes/functor.ml (hunk 14 above)
Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml:516]: Yeah we allow this because it's sound. I'm think forbidding this might break some existing code.

File: testsuite/tests/typing-modes/functor.ml (hunk 17 above)
Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml:320]: Added. 
FWIW the content of mode axes are orthoganol to the PR; if the tests failed we have a bigger fundamental problem.

File: testsuite/tests/typing-modes/functor.ml
Comment (riaqn) [reply testsuite/tests/typing-modes/functor.ml:1]: Added.
FWIW the difference between monadic and comonadic axes is orthoganol to (and not modified in) this PR.
